provider/equinix/environ.go

```diff
@@ -0,0 +1,649 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package equinix
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/juju/cloudconfig/cloudinit"
+	"github.com/juju/juju/cloudconfig/instancecfg"
+	"github.com/juju/juju/cloudconfig/providerinit"
+	"github.com/juju/juju/core/constraints"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/environs"
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/imagemetadata"
+	"github.com/juju/juju/environs/instances"
+	envinstance "github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/environs/simplestreams"
+	"github.com/juju/juju/provider/common"
+	"github.com/juju/juju/storage"
+	"github.com/juju/juju/tools"
+	"github.com/juju/os/series"
+	"github.com/juju/retry"
+	"github.com/juju/schema"
+	"github.com/juju/utils/arch"
+	"github.com/juju/version"
+	"github.com/lxc/lxd/shared/logger"
+	"github.com/packethost/packngo"
+	"gopkg.in/juju/environschema.v1"
+)
+
+type environConfig struct {
+	config *config.Config
+	attrs  map[string]interface{}
+}
+
+type environ struct {
+	ecfgMutex    sync.Mutex
+	ecfg         *environConfig
+	name         string
+	cloud        environscloudspec.CloudSpec
+	equnixClient *packngo.Client
+}
+
+var providerInstance environProvider
+
+func (e *environ) AdoptResources(ctx context.ProviderCallContext, controllerUUID string, fromVersion version.Number) error {
+	return nil
+}
+
+func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.ProviderCallContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tags map[string]string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	packetTags := []string{}
+	for k, v := range tags {
+		packetTags = append(packetTags, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	devices, _, err := e.equnixClient.Devices.List(e.cloud.Credential.Attributes()["project-id"], nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, d := range devices {
+		cp := d
+		if isListContained(packetTags, cp) {
+			toReturn = append(toReturn, &equnixDevice{e, &cp})
+		}
+	}
+
+	return toReturn, nil
+}
+
+func isListContained(tags []string, d packngo.Device) bool {
```

Comment:
Please use `set.Strings` as suggested in the WIP PR instead of custom function. You can create two sets and check if `a.Intersection(B).IsEmpty()`

Reply:
done